src/libraries/System.Private.CoreLib/src/System/Reflection/UninitializedObjectFactory.cs

```diff
@@ -0,0 +1,132 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Reflection
+{
+    public abstract class UninitializedObjectFactory
```

Comment:
This wasn't intended specifically for binary-serialization like schemes. There are two scenarios that I think could benefit from it.

 1. Even in serialization schemes like JSON, instantiating an object during deserialization is still a two-step process: (a) allocate the memory, then (b) call the appropriate ctor. In this PR I support the most basic scenario of a public parameterless ctor. However, there might be other ctor patterns that _System.Text.Json_ or other serialization libraries want to support in the future. By exposing `UninitializedObjectFactory` as a first-class concept, they can quickly create a zero-inited instance of the target type, then use their own logic to call a cached ctor delegate over the newly-allocated instance. This allows them an additional layer of flexibility in case we don't natively support the pattern they're trying to match.

 2. Scenarios like DI are a special case of the above. The DI framework might natively understand some concepts like `new T(IServiceProvider)` that we don't support out-of-box. Assuming they're able to build a ctor delegate cache, they can use `UninitializedObjectFactory` to quickly create a "blank" object, then turn around and immediately feed that object into the cached ctor delegate. This pattern should rival most hand-rolled `DynamicMethod` schemes in terms of throughput.

I don't see `BinaryFormatter` or other such components making use of this API. Those components don't have much ability to cache because the type information is determined by the payload itself. They're best served by the existing slower API surface. Contrast this against _System.Text.Json_ and similar frameworks which are expected to be called many times over atop a known destination type. These new APIs are better suited for those scenarios.

Reply:
Agreed with @jkotas. It's possible to deserialize an object from any unordered data (am working on it). Two weeks ago I started my research in this area for Npgsql and did some progress. The idea can be easily reused by other serializers.

In case when an object has a parametrized constructor there must be a temporary storage for arguments. If there are storage it can be used even to hold data which should be moved to an object after it construction.

Since the temporary storage should be constructed at runtime a type builder should be used and not `System.Reflection.Emit` because there is a requirement by AOT. Do not forget that the is no way to construct generics on the fly, but any well designed serializer consists of type handlers which also named as value converters. Converters are strongly typed (in case of the JSON serializer `JsonConverter<T>`). Therefore, they can report or construct a storage for a member.

```csharp
public abstract class JsonConverter
{
    internal abstract StorageField CreateStorage(StorageFactory factory);
}

public abstract class JsonConverter<T>
{
    internal override StorageField CreateStorage(StorageFactory factory) =>
        factory.AddField<T>();
}

internal readonly struct StorageField
{
    public int Index { get; }
    public StorageField(int index) => Index = index;
}

internal readonly struct Storage
{
    private readonly IStorageInternal;

    public T GetValue<T>(StorageField field);
    public void SetValue<T>(StorageField field);
}

internal sealed class StorageFactory
{
    public StorageField AddField<T>();
    public Storage CreateStorage();
}
```

`StorageFactory` implementation can be the same as I did for `TupleHandler` in my research project (https://github.com/YohDeadfall/Yoh.Tuples). Internally it has an immutable builder which is generic and is able to produce an actual storage allocated as a single block (a boxed structure).

As said above the storage should be used until an object isn't constructed yet, so we need a member handler like this:

```csharp
internal abstract class JsonPropertyHandler<TClass>
{
    public abstract void Read(ref Utf8JsonReader reader, TClass instance, Storage storage, JsonSerializerOptions options);

    public abstract void Move(TClass instance, Storage storage);
}

internal sealed class JsonPropertyHandler<TClass, TMember> : JsonPropertyHandler
{
    private readonly JsonConverter<TMember> _converter;
    private readonly Action<TClass, TMember> _setter;
    private readonly StorageField _field;

    public override void Read(ref Utf8JsonReader reader, TClass instance, Storage storage, JsonSerializerOptions options)
    {
        var value = _converter.Read(ref reader, typeof(TMember), options);
        if (instance is null)
            _storage.SetValue(_field, value);
        else
            _setter(instance, value);
    }

    public override void Move(TClass instance, Storage storage) =>
        _setter(instance, storage.GetValue<TMember>(_field));
}
```

Just a note. Instead of passing an instance directly there can be used an object (don't like it since requires boxing for structs, but Npgsql uses this way because its handlers are async, see [this for info](https://github.com/npgsql/npgsql/blob/5c027d58b923a3310d1004398b981e45428a1137/src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs#L50-L65)). Also the should be a flag if `TClass` is a struct, so makes sence to introduce a second method which will accept something like `Nullable<T>` but without a constraint on `T`, or a second method.

```csharp
internal sealed class JsonClassConverter<TClass> : JsonConverter<TClass>
{
    private readonly StorageFactory _storageFactory;
    private readonly JsonPropertyHandler<TClass>[] _propertyHandlers;
    private readonly JsonClassFactory<TClass>[] _classFactories;

    public abstract TClass Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        var storage = _storageFactory.CreateStorage();
        var instance = default(TClass);
        var constructed = false;

        // actually if there a default constructor we should use it
        // and set constructed to true

        foreach (var propertyHandler in _propertyHandler)
        {
            propertyHandler.Read(ref reader, instance, storage, options);
            // if isn't constructed check that there is a constructor
            // for which all required parameters were read
            // if so construct an instance and move non ctor args
            // field to the instance
        }

        return instance;
    }
}
```

`JsonClassFactory`'s responsibility is to take a storage, get required values and call `ObjectFactory`.

There are some issues with the proposed design since `JsonClassConverter<TClass>` cannot be constructed via `MakeGeneric`. So it should be changed and the generic parameter should be removed. Maybe some other problems, had no time to finish it and test.

---

@steveharter, for your info I contacted the team (@ahsonkhan and @layomia) on this, but got no response on the second mail. This was before @GrabYourPitchforks announced his work in Twitter, so liked to discuss your progress and mine ideas via Teams or like so.

Reply:
`[Uninitialized]ObjectFactory` is gone as of the latest iteration.